fish-rust/src/common.rs

```diff
@@ -1,13 +1,1714 @@
+    let mut prev_was_hex_encoded = false;
+    let narrow = wcs2string(input);
+    let mut out = WString::new();
+    for byte in narrow.into_iter() {
+        if (byte & 0x80) == 0 {
+            let c = char::from_u32(u32::from(byte)).unwrap();
+            if c.is_alphanumeric() && (!prev_was_hex_encoded || c.to_digit(16).is_none()) {
+                // ASCII alphanumerics don't need to be encoded.
+                if prev_was_hex_encoded {
+                    out.push('_');
+                    prev_was_hex_encoded = false;
+                }
+                out.push(c);
+                continue;
+            }
+        } else if byte == b'_' {
+            // Underscores are encoded by doubling them.
+            out += "__"L;
+            prev_was_hex_encoded = false;
+            continue;
+        }
+        // All other chars need to have their UTF-8 representation encoded in hex.
+        out += &sprintf!("_%02X"L, byte)[..];
+        prev_was_hex_encoded = true;
+    }
+    out
+}
+
+/// Escapes a string for use in a regex string. Not safe for use with `eval` as only
+/// characters reserved by PCRE2 are escaped.
+/// \param in is the raw string to be searched for literally when substituted in a PCRE2 expression.
+fn escape_string_pcre2(input: &wstr) -> WString {
+    let mut out = WString::new();
+    out.reserve(
+        (f64::from(u32::try_from(input.len()).unwrap()) * 1.3) // a wild guess
+            .to_i128()
+            .unwrap()
+            .try_into()
+            .unwrap(),
+    );
+
+    for c in input.chars() {
+        if [
+            '.', '^', '$', '*', '+', '(', ')', '?', '[', '{', '}', '\\', '|',
+            // these two only *need* to be escaped within a character class, and technically it
+            // makes no sense to ever use process substitution output to compose a character class,
+            // but...
+            '-', ']',
+        ]
+        .contains(&c)
+        {
+            out.push('\\');
+        }
+        out.push(c);
+    }
+
+    out
+}
+
+/// Escape a string so that it may be inserted into a double-quoted string.
+/// This permits ownership transfer.
+pub fn escape_string_for_double_quotes(input: &wstr) -> WString {
+    // We need to escape backslashes, double quotes, and dollars only.
+    let mut result = input.to_owned();
+    let mut idx = result.len();
+    while idx > 0 {
+        idx -= 1;
+        if ['\\', '$', '"'].contains(&result.char_at(idx)) {
+            result.insert(idx, '\\');
+        }
+    }
+    result
+}
+
+pub fn unescape_string(input: &wstr, style: UnescapeStringStyle) -> Option<WString> {
+    match style {
+        UnescapeStringStyle::Script(flags) => unescape_string_internal(input, flags),
+        UnescapeStringStyle::Url => unescape_string_url(input),
+        UnescapeStringStyle::Var => unescape_string_var(input),
+    }
+}
+
+pub fn unescape_string_in_place(s: &mut WString, style: UnescapeStringStyle) -> bool {
```

Comment:
Can you add a comment:

```rust
/// TODO: Make this actually in-place or rename it to make it clear that it's not an in-place unescape.
```

Reply:
actually the comment should say we plan to remove it